src/vue/modules/assets/asset-explorer/components/asset-actions-bar.spec.js

```diff
@@ -0,0 +1,135 @@
+import AssetActionsBar from './asset-actions-bar'
+
+import Vuex from 'vuex'
+
+import { assetExplorerModule } from '../store/index'
+import { Asset } from '../../shared/wrappers/asset'
+
+import { createLocalVue, shallowMount } from '@vue/test-utils'
+
+import { Bus } from '@/js/helpers/event-bus'
+import { ErrorHandler } from '@/js/helpers/error-handler'
+
+const localVue = createLocalVue()
+localVue.use(Vuex)
+
+describe('Asset explorer module', () => {
+  const props = {
+    asset: new Asset({
+      id: 'USD',
+    }),
+  }
+
+  let store
+  let wrapper
+
+  beforeEach(() => {
+    store = new Vuex.Store({
+      modules: { 'asset-explorer': assetExplorerModule },
+    })
+
+    wrapper = shallowMount(AssetActionsBar, {
+      store,
+      localVue,
+      propsData: props,
+    })
+  })
+
+  describe('method', () => {
+    describe('addBalance', () => {
+      beforeEach(() => {
+        sinon.stub(wrapper.vm, 'createBalance').resolves()
+        sinon.stub(Bus, 'success')
+        sinon.stub(ErrorHandler, 'process')
+      })
+
+      afterEach(() => {
+        wrapper.vm.createBalance.restore()
+        Bus.success.restore()
+        ErrorHandler.process.restore()
+      })
+
+      it('calls Bus.error if balance creation is not allowed', async () => {
+        const kycRequiredAssetType = 'KYC_REQUIRED'
+        wrapper.setProps({
+          kycRequiredAssetType,
+          asset: new Asset({
+            type: kycRequiredAssetType,
+          }),
+          isAccountUnverified: true,
+        })
+        sinon.stub(Bus, 'error')
+
+        await wrapper.vm.addBalance()
+
+        expect(Bus.error).to.have.been.calledOnce
+
+        Bus.error.restore()
+      })
+
+      it('calls createBalance method with correct params', async () => {
+        sinon.stub(wrapper.vm, 'loadAccountBalances').resolves()
+
+        await wrapper.vm.addBalance()
+
+        expect(wrapper.vm.createBalance)
+          .to.have.been.calledOnceWithExactly('USD')
+
+        wrapper.vm.loadAccountBalances.restore()
+      })
+
+      it('calls loadAccountBalances method', async () => {
```

Comment:
overtested. a function should perform an action, you should check that the action performed correctly.

all the prop changes, all the function calls, and all the other things related to the action should be checked in the same test. otherwise you cannot say what actually the test asserts.

it is not unit testing anymore, it is unit-unit testing now.


But it is actually a nice hint: if you feel like needed to assert every effect in separate test it is time to start suspecting you did something wrong about this function

Reply:
Updated

Reply:
It will look better if you write such tests in this way:
```
describe('myFunction', () => {
  before(() => {
    sinon.stub()
    sinon.stub()
    ...
    myFunction()
  })

  it('should do smth') // only assertion inside
  it('should do smth else')
  ...
  it('')

  after(() => {
    sinon.restore()
  })
})
```

it's easy to define what exactly failed, because the messages will be more declarative. and your test will be kinda optimised cause you won't repeat the stubs and calls in every `it` call